zebra/zebra_nhg.c

```diff
@@ -177,10 +177,16 @@ static int nexthop_active(afi_t afi, struct route_entry *re,
 	re->nexthop_mtu = 0;
 
 	/*
-	 * If the kernel has sent us a route, then
+	 * If the kernel has sent us a NEW route, then
 	 * by golly gee whiz it's a good route.
+	 *
+	 * If its an already INSTALLED route we have already handled, then the
+	 * kernel route's nexthop might have became unreachable
+	 * and we have to handle that.
 	 */
-	if (re->type == ZEBRA_ROUTE_KERNEL || re->type == ZEBRA_ROUTE_SYSTEM)
+	if (!CHECK_FLAG(re->status, ROUTE_ENTRY_INSTALLED)
```

Comment:
The new comment and the new test aren't quite coming together for me? the comment says "if it's [sic] an INSTALLED route", but then the test in the `if` is for NOT INSTALLED.

Reply:
If this route has never been installed before and its a kernel route we must have JUST gotten it from a kernel message (i.e. NEW). Trust then that its all good, because the kernel would not send us a bad one.

If this routes is already installed (i.e. something internally told us to process it) (i.e. OLD) then we need to check its nexthop is still good.

so the 
```
if (!CHECK_FLAG(re->status, ROUTE_ENTRY_INSTALLED)
```
is determining whether this route is NEW or OLD. If its never been installed, it must be NEW and we can trust it. Otherwise, its OLD and we need to reprocess it.

Reply:
wow - that's quite something! ok, I get it...